File: gametimer.cpp
#include "gametimer.hpp"

GameTimer::GameTimer() : elapsedSeconds(0), isRunning(false)
{
    // Обновляем время после каждой пройденной секунды //
    connect(&timer, &QTimer::timeout, this, &GameTimer::updateTimer);

}

void GameTimer::start() {

    if (!isRunning) {

        isRunning = true;
        timer.start(1000); // запускаем таймер на 1 секунду

    }
}

void GameTimer::stop() {

    if (isRunning) {

        isRunning = false;
        timer.stop();

    }
}

void GameTimer::restart() {

    stop();
    elapsedSeconds = 0;

}

int GameTimer::getElapsedSeconds() const {
    return elapsedSeconds;
}

bool GameTimer::getIsRunning() const {
    return isRunning;
}

QString GameTimer::getFormattedTime() const {

    int minutes = elapsedSeconds / 60;
    int seconds = elapsedSeconds % 60;

    // Форматируем время как MM:SS //
    return QString("%1:%2").arg(minutes, 2, 10, QChar('0')).arg(seconds, 2, 10, QChar('0'));

}

void GameTimer::updateTimer() {

    if (isRunning) {
        elapsedSeconds++;
        emit timeUpdated(getFormattedTime());
    }

}

File: GameTimer.cs
using System;
using System.Windows.Forms; // Таймер

namespace Lab3
{
    public class GameTimer
    {

        private const int TIMER_INTERVAL_MS = 1000; // Интервал обновления таймера в 1000 мс
        private const int SECONDS_PER_MINUTE = 60; // 60 секунд = 1 минута

        private Timer _timer;

        public int ElapsedSeconds { get; set; }
        public bool IsRunning { get; set; }

        public event Action<string> TimeUpdated; // Событие (аналог сигналам и слотам в Qt)

        public GameTimer()
        {

            _timer = new Timer();
            _timer.Interval = TIMER_INTERVAL_MS;
            _timer.Tick += Timer_Tick; // Подписываемся на событие Tick (аналог QTimer::timeout)

        }

        public void Start()
        {

            if (!IsRunning)
            {

                IsRunning = true;
                _timer.Start();

            }

        }

        public void Stop()
        {

            if (IsRunning)
            {

                IsRunning = false;
                _timer.Stop();

            }

        }

        public void Restart()
        {

            Stop();
            ElapsedSeconds = 0;
            TimeUpdated?.Invoke(GetFormattedTime());

        }

        public string GetFormattedTime()
        {

            int minutes = ElapsedSeconds / SECONDS_PER_MINUTE;
            int seconds = ElapsedSeconds % SECONDS_PER_MINUTE;


            return $" {minutes:D2}:{seconds:D2}";
        }

        public void Timer_Tick(object sender, EventArgs e)
        {

            if (IsRunning)
            {

                ElapsedSeconds++;
                TimeUpdated?.Invoke(GetFormattedTime());

            }

        }

        // Очистка ресурсов таймера //
        public void Dispose()
        {

            _timer.Stop();
            _timer.Tick -= Timer_Tick; // Отписываемся от события
            _timer.Dispose(); // Освобождаем ресурсы таймера

        }

    }
}
